### DISCUSSION TITLE
Are SCENIC+ results stochastic?

### DISCUSSION THREAD

**Discussion by liz-is** [aertslab/scenicplus#39] [Q&A] [answered]
Hello, I just have a quick request for clarification. I know that SCENIC has some stochasticity, such that different runs will return slightly different regulons/target gene sets. Is the same true for SCENIC+? 

I was expecting that this would be the case, but my initial inspection of a small subset of the results of two independent runs of SCENIC+ on the same data (using the ` run_scenicplus` wrapper) suggests that the target genes of the regulons and the various metadata column values are the same between these two runs. So I wanted to check if this is expected, or if there's an issue with my analysis! 

Thank you!

**Comment by SeppeDeWinter** [comment, marked as answer]
Hi @liz-is 

I moved your issue to discussion because it is more a question than an issue with the code.

It is true that SCENIC+ has some stochasticity (for example the gradient boosting machine we use for calculating region to gene links and TF to gene links is not completely deterministic). 
The reason why you get the same result is because all of the function which have some stochasticity (randomness) are seeded. Meaning that if you don't change the seed and all other parameters are kept the same (exact same input data etc.) your output should indeed not change.

Hope this clarifies things.

Best,

Seppe

**Reply by liz-is** [reply]
Indeed, this is a better place for it, thanks! (I'm not used to using the Discussions tab yet :) )

I figured that it could well be the case that there's a seed that's set, that makes sense. Can you provide any tips on how I would go about changing the seed? I don't see an option to set a seed as a named argument to `run_scenicplus`, although I see that this is an option in e.g. `generate_pseudobulks`. The **kwargs for `run_scenicplus` seem to be passed down to `ray.init()` - I have no experience with ray, unfortunately, so I get lost here. 

Apologies if I'm missing something obvious. Maybe there's a standard pythonic way to do this I'm not aware of yet!

**Reply by SeppeDeWinter** [reply]
We are also getting used to this tab, so no worries ;)

No need to apologise. At this moment the seed is hardcoded, so it's not possible to change it by setting a parameter. We should actually change this. 
Thank you for the comment, we will add an argument for setting the seed in a future update.

Best,

Seppe

**Reply by liz-is** [reply]
Okay, got it! I look forward to the update it :)

**Reply by liz-is** [reply]
Hello,

First, congrats on the publication of SCENIC+! :)

I just wanted to follow up and see if you are still planning to introduce an argument to change the seed for SCENIC+ runs. (I had a look through recent commits and I don't think I saw this feature added, but apologies if I missed it!)

**Reply by SeppeDeWinter** [reply]
@liz-is 

Thank you.

I'm currently working on refactoring the SCENIC+ code (see development branch, https://github.com/aertslab/scenicplus/tree/development). There I have added the seed as an argument to each function. The next version of SCENIC+ will use Snakemake instead of the run_scenicplus wrapper, with this you will be able to set the seed (see https://github.com/aertslab/scenicplus/blob/development/Snakemake/config/config.yaml).

Best,

Seppe
